### ISSUE TITLE
[Camera] camera.takePictureAsync() responds only after second call

### ISSUE THREAD

**Issue by Vieshanis** [expo/expo#2288]
The issue is that I receive response from takePictureAsync only after second press on button

Here is my code

```
takePicture() {
    this.setState({
        takeImageText: "... PROCESSING PICTURE ..."
    });
    this.camera.takePictureAsync({ skipProcessing: true }).then((data) => {
        this.setState({
            takeImageText: "PICTURE TAKEN",
            photo: data.uri
        }, console.log(data.uri))
    })
}

<Camera
    ref={ref => { this.camera = ref; }}
    type={this.state.type} >
    <View
        style={{
            flex: 1,
            backgroundColor: 'transparent',
            flexDirection: 'row',
        }}>
        <TouchableOpacity
            onPress={() => {
                this.setState({
                    type: this.state.type === Camera.Constants.Type.back
                        ? Camera.Constants.Type.front
                        : Camera.Constants.Type.back,
                });
            }}>
            <Text>
                Flip
            </Text>
        </TouchableOpacity>
    </View>
</Camera>
<View>
    <TouchableOpacity
        onPress={this.takePicture.bind(this)} >
        <Text>Take photo</Text>
    </TouchableOpacity>
</View>
```
package.json
```
  "dependencies": {
    "@expo/samples": "2.1.1",
    "expo": "^30.0.0",
    "react": "16.3.1",
    "react-native": "https://github.com/expo/react-native/archive/sdk-30.0.0.tar.gz",
    "react-native-camera": "^1.2.0",
    "react-native-datepicker": "^1.7.2",
    "react-native-keyboard-aware-scroll-view": "^0.7.2",
    "react-native-segmented-control-tab": "^3.3.1",
    "react-native-svg": "^6.5.2",
    "react-native-svg-charts": "^5.2.0",
    "react-native-tab-view": "^1.0.2",
    "react-navigation": "^2.9.3",
    "whatwg-fetch": "^2.0.4"
  },
  "devDependencies": {
    "jest-expo": "30.0.0"
  }
```
app.json
`"sdkVersion": "30.0.0",`

**Comment by bami** [comment]
Hi, I've the same issue. How did you solve it?

**Comment by Vieshanis** [comment]
@bami Simply removed first setState from func

 ```
   takePicture() {
        this.camera.takePictureAsync({ skipProcessing: true }).then((data) => {
            this.setState({
                photo: data.uri
            });
        });
    }
```

**Comment by bami** [comment]
hm. reducing the load in the callback did help me. I only call an async function there. 
downgrading the project to expo 27.0.1 helped. 

_package.json_
`
"dependencies": {
    "expo": "^27.0.1",
    "react": "16.3.1",
    "react-native": "~0.55.2",
    "react-navigation": "^2.16.0"
`

and _app.json_
`
"sdkVersion": "27.0.0",
`

**Comment by levansuper** [comment]
an ugly but working solution 


`
 snap = async () => {
    setTimeout(() => {
      this.setState({
        isTakingImage: true
      })
    }, 1);
    const image = await this.camera.takePictureAsync();
      this.props.onImage(image);
      this.setState({
        isTakingImage: false
      })
  };

`

**Comment by bami** [comment]
@Vieshanis Vieshanis

Thanks for you approach. As far as I understand, the difference is, 
1) camera is not checked
2) that the call inside processing the result of taking the picture inside .then(...) is executed synchronously instead of asynchronously.

Maybe 1) speeds up the processing a very tiny bit. But I think 2) does not make any difference, since the calls are async anyway. If the processing would get blocked by the synchronously, the GUI update would even get slower. Don't you agree?!

**Comment by burivuhster** [comment]
I have the same issue on iOS. And it only reproduces in a production build (via TestFlight). 
I'm using the latest SDK v31.

The issue occurs even 10 seconds after opening the camera screen. So seems it's not due to some long camera initialization.

I've noticed an interesting pattern: 
first tap on "capture" button – nothing happens (first  takePictureAsync call)
second tap - I hear **two** camera shutter sounds and then it captures a picture
BUT
if I make some **other** action after the first try (for example switch camera to front, or **even just press "Back" button** on the navbar) - I hear this camera shutter sound and the callback is finally fired.

**Comment by muZk** [comment]
I'm using v30 and I'm facing the same problem that @burivuhster described.

**Comment by muZk** [comment]
My current solution is to use `onPictureSaved` callback and `setState` just after `takePictureAsync`:

```
this.camera.takePictureAsync({
   onPictureSaved: this.handlePictureSaved,
});
this.setState({ cameraDisabled: true });
```

I think that `await` solution won't work because `takePictureAsync` isn't finishing its execution and it won't execute the next line (as @burivuhster said).

**Comment by burivuhster** [comment]
@muZk thanks for your advice, 
I added `this.setState({ foo: Math.random() });` after takePictureAsync and somehow it helped.
My takePicture function looks like that now:
```
this.camera.current.takePictureAsync({
  onPictureSaved: this.onPictureSaved,
});
this.setState({
  foo: Math.random()
});
```
@Vieshanis please consider re-opening this issue. Our setState solution obviously is not a fix of an actual underlying problem with camera component.

**Comment by asgerhallas** [comment]
I'm having the same problem. Inserting a console.log('') just before `await this.camera.current.takePictureAsync()` also "fixes" the problem :)

**Comment by jurs94** [comment]
The solution of @asgerhallas works, but (at least for me) only in development.

**Comment by jurs94** [comment]
The solution of @asgerhallas only works in development

**Comment by supreme** [comment]
Still present in SDK v31

**Comment by asgerhallas** [comment]
@supreme Odd. I don't experience the problem in SDK v31 on iOS... 
@jurs94 yes, you're right, didn't think of that. Never got it that far :)

**Comment by stewartduffy** [comment]
Issue is still present in SDK v31. iOS only in prod

**Comment by tozaicevas** [comment]
> I'm having the same problem on iOS. Inserting a console.log('') just before `await this.camera.current.takePictureAsync()` also "fixes" the problem :)

Im having a quite similar issue, however I take use of onPictureSaved callback and send the base64 to external API. It takes a picture but the response from the API doesn't come back, even tho the api is working and is well tested. However, if I put a console.log(""); in fetch().then(response => console.log(""))); it works perfectly

**Comment by tozaicevas** [comment]
What exactly doesnt work for you in this snack? Seems like its working fine for me.

**Comment by AlbuquerqueRafael** [comment]
I'm having the same problem. When I use _this.setState_ before takePictureAsync, takePictureAsync's promise does not resolve. When I remove _this.setState_ from my function, takePictureAsync works like a charm.

**Comment by tozaicevas** [comment]
Try writing `setTimeout(() => this.forceUpdate(), 1);` after takePictureAsync. This workaround has worked for me after publishing.

**Comment by serhiipalash** [comment]
```
setTimeout(() => this.forceUpdate(), 1);
```
It doesn't work for us, the bug is still there. Device tested is iPhone 6s.

**Comment by mrcarjul** [comment]
This is the only way i could make it work in prod(tested in iphone 5s and iphone 8),
i don't use an async await cycle in takePictureAsync(otherwise it gets stuck there forever), and i change the state after that with a timeout (and make sure that this state modifies something in the screen, in my case it was just a flex: 0 to a flex: 0. 1 of a View), i really hope this helps.
`  takePicture = () => {
    if (this.camera) {
      try {
        this.setState(
          {
            spinner: true,
            isTakingImage: true
          },
          () =>
            this.camera.takePictureAsync(
              {
                onPictureSaved: picture => this.processPicture(picture),
                base64: true,
                quality: 0.8
              },
              rejected => {
                console.log('Rejected by: ' + rejected);
                this.setState({ recognitionFail: true });
              }
            )
        );
        if (Platform.OS !== 'android') {
          // Workaround for iOS
          clearTimeout(checkProccess);
          const checkProccess = setTimeout(() => {
            if (this.state.spinner && this.state.isTakingImage) {
              this.setState({
                isTakingImage: false
              });
            }
          }, 1500);
        }
      } catch (err) {
        console.log(err);
        this.setState({ recognitionFail: true, isTakingImage: false });
      }
    }
  };`

**Comment by serhiipalash** [comment]
It works! Thank you @mrcarjul !

**Comment by CaesarXing** [comment]
any updates？

**Comment by aarjae** [comment]
> @muZk thanks for your advice,
> I added `this.setState({ foo: Math.random() });` after takePictureAsync and somehow it helped.
> My takePicture function looks like that now:
> 
> ```
> this.camera.current.takePictureAsync({
>   onPictureSaved: this.onPictureSaved,
> });
> this.setState({
>   foo: Math.random()
> });
> ```
> @Vieshanis please consider re-opening this issue. Our setState solution obviously is not a fix of an actual underlying problem with camera component.

How do you get access to the picture object in your this.onPictureSaved function?

**Comment by logicallyDimp** [comment]
apparently this works for me
```js
this.camera.takePictureAsync({ quality: 1, skipProcessing: false }).then((data) => {

      this.setState({
        image: data.uri,
      });

    });
    this.setState({
      openCamera: false,
    });
  }
```

**Comment by tacomanator** [comment]
I found the issue to still exist in SDK32.

Thank you everyone for sharing your workarounds. Unfortunately most work only inconsistently. I tested many variations (on iPhone X) finally settled on the following. The only one that worked every time was @levansuper's solution:

```javascript
takePicture = async () => {
  setTimeout(() => this.setState({ isTakingImage: true }), 1);
  this.camera.takePictureAsync().then(image => {
    this.setState({ isTakingImage: false });
    this.onReceiveImage(image);
  });
  if (Constants.isDevice) this.camera.pausePreview();
}
```

Or, if you don't need to pause the preview:

```javascript
takePicture = async () => {
  setTimeout(() => this.setState({ isTakingImage: true }), 1);
  const image = await this.camera.takePictureAsync();
  this.onReceiveImage(image);
  this.setState({ isTakingImage: false });
}
```

**Comment by nandorojo** [comment]
This issue is still very much unsolved in v32. Photos take a random amount of time to resolve, and many times they don't resolve at all. I've found a work-around of my own I can share, but it's very messy and isn't in the expected behavior. I'm happy to upload a snack if people still need it after the proposed workarounds above.

Here's a snack of an example of the expo camera: https://snack.expo.io/@git/github.com/nandorojo/expo-camera-bug-v32

Thanks!

**Comment by dlerman2** [comment]
Confirmed unresolved on v31 and v32, and the same problem exists for `camera.recordAsync`.

I'm currently working around it by preceding every call to `camera.takePictureAsync`, `camera.recordAsync` and `camera.stopRecording` with:

```
if (Platform.OS !== 'android') {
  setTimeout(() => this.setState({ __hackCameraFix: uuid() }), 1);
}
```

**Comment by dlerman2** [comment]
Never mind, the setTimeout hack appeared to fix it, but not 100% of the time. Now trying the following, but I'm not sure if this actually fixes it by defering execution until the following frame so the native code has a chance to run, or if it just seems to fix it by waiting a little longer to call setState. This desperately needs a real solution, but I unfortunately don't know the native side enough to dig in...

```
if (Platform.OS !== 'android') {
  requestAnimationFrame(() => {
    requestAnimationFrame(() => component.setState({ __hackCameraFix: uuid() }));
  });
}
```

**Comment by mauriciord** [comment]
@tacomanator How can i apply your solution to the [expo example using Camera](https://github.com/expo/camerja/blob/master/App.js#L118)

I'm using sdk v.32.0.0 currently and this problem persists

**Comment by mauriciord** [comment]
@brentvatne @ide

**Comment by aarjae** [comment]
@logicallyDimp skipping processing has undesirable side effects on cross platforms

**Comment by epurban** [comment]
This is a huge bug. How is this not fixed yet???

**Comment by yangga** [comment]
I have a same problem. Is it fixed?

```
...
  "dependencies": {
    "@expo/vector-icons": "^10.0.1",
    "autobind-decorator": "^2.4.0",
    "crypto-js": "^3.1.9-1",
    "expo-asset": "^4.0.0",
    "expo-blur": "^4.0.0",
    "expo-camera": "^4.0.0",
    "expo-font": "^4.0.0",
    "expo-gl": "^4.0.0",
    "expo-gl-cpp": "^4.0.0",
    "expo-linear-gradient": "^4.0.0",
    "lodash": "^4.17.11",
    "mobx": "4.9.2",
    "mobx-react": "5.4.3",
    "moment": "^2.24.0",
    "prop-types": "^15.7.2",
    "react": "16.8.6",
    "react-native": "0.59.8",
    "react-native-datepicker": "^1.7.2",
    "react-native-gesture-handler": "^1.2.1",
    "react-native-keyboard-spacer": "^0.4.1",
    "react-native-router-flux": "^4.0.6",
    "react-native-svg": "^9.4.0",
    "react-native-unimodules": "0.3.1",
    "react-navigation": "^3.11.0"
  },
...
```

**Comment by sjchmiela** [comment]
I think you should be able to upgrade `react-native-unimodules` to `0.4.0-alpha.3` (you'll need to change your `AppDelegate.m` and remove `andExperience:nil` from:

https://github.com/expo/expo/blob/master/templates/expo-template-bare-minimum/ios/HelloWorld/AppDelegate.m#L41

**Comment by yangga** [comment]
> I think you should be able to upgrade `react-native-unimodules` to `0.4.0-alpha.3` (you'll need to change your `AppDelegate.m` and remove `andExperience:nil` from:
> 
> https://github.com/expo/expo/blob/master/templates/expo-template-bare-minimum/ios/HelloWorld/AppDelegate.m#L41

Unfortunately, it's not working for me. :(

**Comment by kishorekondepudi** [comment]
I had the same issue in my application. Basically my app has a pull over drawer overlayed on Camera. initially i had fixed height for camera so that it ends at the place where the bottom drawer begins.

My UI

![IMG_0037](https://user-images.githubusercontent.com/18567180/58984846-e5156080-878e-11e9-9c71-12f2596ac8fe.jpg)

And i noticed that it never responds on first click and also after the first click if I click anywhere underneath the touchable button it triggered the second click on snap which was weird ideally it should have responded only on the button. How ever any where above the button the camera doesn't respond.

I suspected this to be with the view layout and made my drawer opacity to 0.1 or so to see where my camera view was ending and realized it was due to the overlap zone.

To eliminate the view issue I removed all and only rendered the Camera and it worked absolutely fine with async componentWillMount and after further review I fixed the layout issue and then it started working fine from the first click too.

This may not bee every ones scenario but it will be easier to remove all other view and only have the camera and see if that works. That probably means something else on view layer is responding , in that case opacity might be helpful to deebug.

Just though to share my experience with the issue in case it helps some one.

**Comment by therustydev** [comment]
I am experimenting with taking screenshots of the viewfinder instead of taking pictures with the camera itself. I will give feedback once I am done

**Comment by AndrewSouthpaw** [comment]
FYI I used @tacomanator but needed to wrap the `this.setState({ isTakingImage: false })` in a `setTimeout` as well because of a race condition running in my simulator where it would set the `isTakingImage` to `false` before it was set to `true` 😂 

So mine looks like this:

```js
  toggleTakingPhotoState = (isTakingPicture: boolean) => {
    setTimeout(() => { this.setState({ isTakingPicture }) }, 1)
  }

  takePicture = async () => {
    this.toggleTakingPhotoState(true)
    const image = await this.camera.takePictureAsync()
    this.onReceiveImage(image)
    this.toggleTakingPhotoState(false)
  }

```

**Comment by churichard** [comment]
I'm running into a similar issue on SDK 36, and @tacomanator's solution doesn't work for me. On iOS, the promise that's returned by `this.camera.recordAsync` does not resolve, even after recording has been stopped. This issue is not present on Android.

**Comment by DevelopSmith** [comment]
For me, it gets stuck when it reaches this line (on iPhone via TestFlight):
`const data = await camera.current.takePictureAsync(options)`

**Comment by ShivanshJ** [comment]
It's sad how there is no fix for the delay after clicking the picture too. I am using expo-camera ~8.3.1 and can't seem to get it to work either.

**Comment by brentvatne** [comment]
@ShivanshJ @DevelopSmith - can either of you please create a new issue and fill out the issue template accordingly, in particular we would like to have a reproducible example. thanks!
